File: roman-to-integer/Solution13.java
// https://leetcode.com/problems/roman-to-integer/

import java.util.Map;

class Solution13 {

    private static final Map<Character, Integer> SYMS = Map.of(
            'I', 1,
            'V', 5,
            'X', 10,
            'L', 50,
            'C', 100,
            'D', 500,
            'M', 1000);

    // T: O(n), S: O(1)
    public int romanToInt(String s) {
        int n = s.length(), ans = 0, i = 0;
        while (i < n) {
            int currVal = SYMS.get(s.charAt(i));
            int nextVal = (i < n - 1) ? SYMS.get(s.charAt(i + 1)) : 0;
            if (currVal < nextVal) {
                ans += (nextVal - currVal);
                i += 2;
            } else {
                ans += currVal;
                i++;
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        var sol = new Solution13();
        var s1 = "III";
        var s2 = "LVIII";
        var s3 = "MCMXCIV";
        System.out.println(sol.romanToInt(s1)); // 3
        System.out.println(sol.romanToInt(s2)); // 58
        System.out.println(sol.romanToInt(s3)); // 1994
    }

}

File: roman-to-integer/solution13.py
# https://leetcode.com/problems/roman-to-integer/

class Solution13:

    SYMS = {
        'I': 1,
        'V': 5,
        'X': 10,
        'L': 50,
        'C': 100,
        'D': 500,
        'M': 1000
    }

    # T: O(n), S: O(1)
    def romanToInt(self, s: str) -> int:
        n, ans, i = len(s), 0, 0
        while i < n:
            curr_val = self.SYMS[s[i]]
            next_val = self.SYMS[s[i + 1]] if i < n - 1 else 0
            if curr_val < next_val:
                ans += (next_val - curr_val)
                i += 2
            else:
                ans += (curr_val)
                i += 1
        return ans


def main() -> None:
    sol = Solution13()
    print(sol.romanToInt("III"))
    print(sol.romanToInt("LVIII"))
    print(sol.romanToInt("MCMXCIV"))


if __name__ == "__main__":
    main()
